File: src/llm_cal/architecture/detector.py
"""`detect()` — main orchestration over trait sub-detectors.

Step 1: Family dispatch (state_space vs transformer vs unknown).
Step 2: Gather traits (independent sub-detectors).
Step 3: Assemble Profile with a confidence level.

Fallback path: `_fallback_unknown()` for configs missing key fields. This is
the bedrock of "works on day-0" — new model types degrade gracefully.
"""

from __future__ import annotations

from typing import Any

from llm_cal.architecture.profile import (
    ArchitectureProfile,
    Confidence,
    Family,
)
from llm_cal.architecture.traits import (
    detect_attention,
    detect_moe,
    detect_position,
    detect_sliding_window,
)

# Model types we know we handle well. Maintained alongside engine_compat matrix.
KNOWN_MODEL_TYPES: frozenset[str] = frozenset(
    {
        "llama",
        "mistral",
        "mixtral",
        "qwen2",
        "qwen2_moe",
        "qwen3",
        "qwen3_moe",
        "deepseek_v2",
        "deepseek_v3",
        "deepseek_v3_2",
        "deepseek_v4",
        "gemma",
        "gemma2",
        "gemma3",
        "phi",
        "phi3",
    }
)

STATE_SPACE_TYPES: frozenset[str] = frozenset({"mamba", "mamba2", "falcon_mamba", "jamba"})


def detect(config: dict[str, Any]) -> ArchitectureProfile:
    """Main entry. Given a parsed config.json dict, return an ArchitectureProfile."""
    model_type = str(config.get("model_type", "")).lower()

    # Step 1: state_space family short-circuits — v0.1 unsupported, but we identify it
    if model_type in STATE_SPACE_TYPES or "ssm_cfg" in config:
        return ArchitectureProfile(
            model_type=model_type,
            architectures=tuple(str(a).lower() for a in config.get("architectures", [])),
            family=Family.STATE_SPACE,
            num_hidden_layers=int(config.get("num_hidden_layers", 0)),
            hidden_size=int(config.get("hidden_size", 0)),
            vocab_size=int(config.get("vocab_size", 0)),
            confidence=Confidence.HIGH,
            auxiliary={"v0_1_unsupported": True},
        )

    # Step 2: reject if fundamentally unidentifiable
    if not model_type and not config.get("architectures"):
        return _fallback_unknown(config)

    # Step 3: required fields
    num_layers = config.get("num_hidden_layers")
    hidden_size = config.get("hidden_size")
    if not num_layers or not hidden_size:
        return _fallback_unknown(config)

    # Step 4: gather traits (each is independent and may return None)
    attention = detect_attention(config)
    moe = detect_moe(config)
    position = detect_position(config)
    sliding = detect_sliding_window(config)

    # Step 5: confidence — HIGH iff model_type is in the registry
    confidence = Confidence.HIGH if model_type in KNOWN_MODEL_TYPES else Confidence.MEDIUM

    # Pass-through of config fields our formulas can use downstream. Keeps the
    # Profile schema stable while enabling richer computation (e.g. dense FFN
    # param count needs intermediate_size).
    auxiliary: dict[str, object] = {}
    if isinstance(config.get("intermediate_size"), int):
        auxiliary["intermediate_size"] = config["intermediate_size"]
    if config.get("tie_word_embeddings") is not None:
        auxiliary["tie_word_embeddings"] = bool(config["tie_word_embeddings"])

    return ArchitectureProfile(
        model_type=model_type,
        architectures=tuple(str(a).lower() for a in config.get("architectures", [])),
        family=Family.TRANSFORMER,
        num_hidden_layers=int(num_layers),
        hidden_size=int(hidden_size),
        vocab_size=int(config.get("vocab_size", 0)),
        confidence=confidence,
        attention=attention,
        moe=moe,
        position=position,
        sliding_window=sliding,
        auxiliary=auxiliary,
    )


def _fallback_unknown(config: dict[str, Any]) -> ArchitectureProfile:
    """Graceful degradation when config.json is unusable.

    Still returns a valid Profile. Consumers check `family == Family.UNKNOWN`
    or `confidence == Confidence.LOW` and skip KV-cache estimation accordingly.
    """
    return ArchitectureProfile(
        model_type=str(config.get("model_type", "")).lower(),
        architectures=tuple(str(a).lower() for a in config.get("architectures", [])),
        family=Family.UNKNOWN,
        num_hidden_layers=int(config.get("num_hidden_layers", 0)),
        hidden_size=int(config.get("hidden_size", 0)),
        vocab_size=int(config.get("vocab_size", 0)),
        confidence=Confidence.LOW,
        auxiliary={
            "warning": (
                "No recognizable model_type or missing essential config fields. "
                "Weight estimate from safetensors file size only; "
                "KV cache cannot be estimated; engine compatibility unknown."
            )
        },
    )

File: crates/llm-infer-cal-core/src/architecture/detector.rs
use std::collections::HashMap;

use serde_json::Value;

use crate::architecture::profile::{ArchitectureProfile, Confidence, Family};
use crate::architecture::traits::{
    detect_attention, detect_moe, detect_position, detect_sliding_window, get, get_truthy_u64,
    get_u64, has_key, truthy,
};

const KNOWN_MODEL_TYPES: &[&str] = &[
    "llama",
    "mistral",
    "mixtral",
    "qwen2",
    "qwen2_moe",
    "qwen3",
    "qwen3_moe",
    "deepseek_v2",
    "deepseek_v3",
    "deepseek_v3_2",
    "deepseek_v4",
    "gemma",
    "gemma2",
    "gemma3",
    "phi",
    "phi3",
];

const STATE_SPACE_TYPES: &[&str] = &["mamba", "mamba2", "falcon_mamba", "jamba"];

pub fn detect(config: &Value) -> ArchitectureProfile {
    let model_type = model_type(config);
    let architectures = architectures(config);

    if STATE_SPACE_TYPES.contains(&model_type.as_str()) || has_key(config, "ssm_cfg") {
        let mut auxiliary = HashMap::new();
        auxiliary.insert("v0_1_unsupported".to_string(), Value::Bool(true));
        return ArchitectureProfile {
            model_type,
            architectures,
            family: Family::StateSpace,
            num_hidden_layers: get_u64(config, "num_hidden_layers").unwrap_or(0),
            hidden_size: get_u64(config, "hidden_size").unwrap_or(0),
            vocab_size: get_u64(config, "vocab_size").unwrap_or(0),
            confidence: Confidence::High,
            auxiliary,
            ..ArchitectureProfile::default()
        };
    }

    if model_type.is_empty() && architectures.is_empty() {
        return fallback_unknown(config);
    }

    let Some(num_hidden_layers) = get_truthy_u64(config, "num_hidden_layers") else {
        return fallback_unknown(config);
    };
    let Some(hidden_size) = get_truthy_u64(config, "hidden_size") else {
        return fallback_unknown(config);
    };

    let attention = detect_attention(config);
    let moe = detect_moe(config);
    let position = detect_position(config);
    let sliding_window = detect_sliding_window(config);
    let confidence = if KNOWN_MODEL_TYPES.contains(&model_type.as_str()) {
        Confidence::High
    } else {
        Confidence::Medium
    };

    let mut auxiliary = HashMap::new();
    let intermediate_size = get(config, "intermediate_size")
        .filter(|value| value.is_i64() || value.is_u64())
        .and_then(crate::architecture::traits::value_to_u64);
    if let Some(intermediate_size) = intermediate_size {
        auxiliary.insert(
            "intermediate_size".to_string(),
            Value::from(intermediate_size),
        );
    }

    let tie_word_embeddings = get(config, "tie_word_embeddings").map(bool_from_value);
    if let Some(tied) = tie_word_embeddings {
        auxiliary.insert("tie_word_embeddings".to_string(), Value::Bool(tied));
    }

    ArchitectureProfile {
        model_type,
        architectures,
        family: Family::Transformer,
        num_hidden_layers,
        hidden_size,
        vocab_size: get_u64(config, "vocab_size").unwrap_or(0),
        confidence,
        attention: Some(attention),
        moe,
        position: Some(position),
        sliding_window,
        intermediate_size,
        tie_word_embeddings: tie_word_embeddings.unwrap_or(false),
        auxiliary,
    }
}

pub fn fallback_unknown(config: &Value) -> ArchitectureProfile {
    let mut auxiliary = HashMap::new();
    auxiliary.insert(
        "warning".to_string(),
        Value::String(
            "No recognizable model_type or missing essential config fields. Weight estimate from safetensors file size only; KV cache cannot be estimated; engine compatibility unknown."
                .to_string(),
        ),
    );

    ArchitectureProfile {
        model_type: model_type(config),
        architectures: architectures(config),
        family: Family::Unknown,
        num_hidden_layers: get_u64(config, "num_hidden_layers").unwrap_or(0),
        hidden_size: get_u64(config, "hidden_size").unwrap_or(0),
        vocab_size: get_u64(config, "vocab_size").unwrap_or(0),
        confidence: Confidence::Low,
        auxiliary,
        ..ArchitectureProfile::default()
    }
}

fn model_type(config: &Value) -> String {
    get(config, "model_type")
        .map(stringify_value)
        .unwrap_or_default()
        .to_lowercase()
}

fn architectures(config: &Value) -> Vec<String> {
    get(config, "architectures")
        .and_then(Value::as_array)
        .map(|values| {
            values
                .iter()
                .map(|value| stringify_value(value).to_lowercase())
                .collect()
        })
        .unwrap_or_default()
}

fn stringify_value(value: &Value) -> String {
    match value {
        Value::String(text) => text.clone(),
        Value::Null => "None".to_string(),
        _ => value.to_string(),
    }
}

fn bool_from_value(value: &Value) -> bool {
    truthy(value)
}
